src/aleph/services/peers/monitor.py

```diff
@@ -51,21 +53,38 @@ async def handle_incoming_host(pubsub_msg: Dict[str, Any], source: Protocol = Pr
 
 
 async def monitor_hosts_p2p(p2p_client: P2PClient) -> None:
-
-    try:
-        stream = await p2p_client.pubsub_subscribe(ALIVE_TOPIC)
-        while True:
-            pubsub_msg = PSMessage()
-            await read_pbmsg_safe(stream, pubsub_msg)
-            msg_dict = pubsub_msg_to_dict(pubsub_msg)
-            await handle_incoming_host(msg_dict, source=Protocol.P2P)
-    except Exception:
-        LOGGER.exception("Exception in pubsub peers monitoring, resubscribing")
+    # The communication with the P2P daemon sometimes fails repeatedly, spamming
+    # IncompleteRead exceptions. We still want to log these to Sentry without sending
+    # thousands of logs.
+    incomplete_read_threshold = 150
+    incomplete_read_counter = 0
+
+    while True:
+        try:
+            stream = await p2p_client.pubsub_subscribe(ALIVE_TOPIC)
+            while True:
+                pubsub_msg = PSMessage()
+                await read_pbmsg_safe(stream, pubsub_msg)
+                msg_dict = pubsub_msg_to_dict(pubsub_msg)
+                await handle_incoming_host(msg_dict, source=Protocol.P2P)
+
+        except IncompleteRead:
+            if (incomplete_read_counter % incomplete_read_threshold) == 0:
+                LOGGER.exception(
+                    "Incomplete read (%d times), reconnecting. Try to restart the application.",
+                    incomplete_read_counter,
+                )
+            incomplete_read_counter += 1
+        except Exception:
+            LOGGER.exception("Exception in pubsub peers monitoring, resubscribing")
+
+        await asyncio.sleep(2)
 
 
 async def monitor_hosts_ipfs(config):
-    try:
-        async for mvalue in sub_ipfs(IPFS_ALIVE_TOPIC):
-            await handle_incoming_host(mvalue, source=Protocol.IPFS)
-    except Exception:
-        LOGGER.exception("Exception in pubsub peers monitoring, resubscribing")
+    while True:
+        try:
+            async for mvalue in sub_ipfs(IPFS_ALIVE_TOPIC):
+                await handle_incoming_host(mvalue, source=Protocol.IPFS)
+        except Exception:
+            LOGGER.exception("Exception in pubsub peers monitoring, resubscribing")
```

Comment:
No sleep here ?

Reply:
I saw no reason to add it here, we do not have this spurious logging issue with IPFS pubsubs so far. I just noticed that the job would die and never be restarted on error.